File: resources/kotoba/shell/app/KotobaShellBridge.java
package {{PACKAGE}};

// kotoba-shell in-app provider bridge — the Android half.
//
// Counterpart to KotobaShellBridge.swift. `bin/kotoba-shell-host-android`
// reaches providers with `adb shell`, which is a developer's machine talking
// to an attached device; a distributed .apk has no adb. These providers are
// compiled into the app.
//
// @JavascriptInterface methods run on a private binder thread and must not
// block, so `request` hands the work to an executor and the result comes back
// through evaluateJavascript into __kotobaShellDeliver. Apple's
// WKScriptMessageHandlerWithReply returns a promise directly; the JS shim
// hides that difference.
//
// The policy decision below matches kotoba.shell.launcher/policy-decision and
// KotobaShellBridge.swift clause for clause. A missing policy asset denies
// everything.

import android.app.Activity;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.Notification;
import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.content.SharedPreferences;
import android.os.Build;
import android.util.Base64;
import android.webkit.JavascriptInterface;
import android.webkit.WebView;

import org.json.JSONObject;

import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileOutputStream;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.security.KeyStore;
import java.util.HashSet;
import java.util.Iterator;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

import javax.crypto.Cipher;
import javax.crypto.KeyGenerator;
import javax.crypto.SecretKey;
import javax.crypto.spec.GCMParameterSpec;

import android.security.keystore.KeyGenParameterSpec;
import android.security.keystore.KeyProperties;

public final class KotobaShellBridge {

    public static final String INTERFACE_NAME = "KotobaShellNative";
    private static final String SCHEMA = "kotoba.shell.bridge.v0";
    private static final String KEY_ALIAS = "kotoba-shell.keychain";
    private static final String KEYCHAIN_PREFS = "{{KEYCHAIN_PREFS}}";
    private static final String DATA_DIR = "{{DATA_DIR}}";
    private static final String NOTIFICATION_CHANNEL = "kotoba-shell";

    private final Activity activity;
    private final WebView webView;
    private final Set<String> allow = new HashSet<>();
    private final Set<String> deny = new HashSet<>();
    private final JSONObject capabilities;
    private final ExecutorService executor = Executors.newFixedThreadPool(2);

    public KotobaShellBridge(Activity activity, WebView webView, JSONObject policy) {
        this.activity = activity;
        this.webView = webView;
        this.capabilities = policy.optJSONObject("capabilities") != null
                ? policy.optJSONObject("capabilities")
                : new JSONObject();
        readTokens(policy, "allow", allow);
        readTokens(policy, "deny", deny);
    }

    private static void readTokens(JSONObject policy, String key, Set<String> into) {
        org.json.JSONArray array = policy.optJSONArray(key);
        if (array == null) {
            return;
        }
        for (int i = 0; i < array.length(); i++) {
            String token = array.optString(i, null);
            if (token != null) {
                into.add(token);
            }
        }
    }

    /**
     * Reads the scaffolded policy asset. A packaging mistake must fail closed:
     * an unreadable policy yields an empty allow set, not an open one.
     */
    public static JSONObject loadPolicy(Context context) {
        try (InputStream in = context.getAssets().open("kotoba-shell-policy.json")) {
            return new JSONObject(new String(readAll(in), StandardCharsets.UTF_8));
        } catch (Exception e) {
            return new JSONObject();
        }
    }

    /** The document-start shim, read from assets so both halves share one source. */
    public static String shimSource(Context context) {
        try (InputStream in = context.getAssets().open("kotoba-shell-bridge.js")) {
            return new String(readAll(in), StandardCharsets.UTF_8);
        } catch (Exception e) {
            return null;
        }
    }

    private static byte[] readAll(InputStream in) throws Exception {
        ByteArrayOutputStream out = new ByteArrayOutputStream();
        byte[] chunk = new byte[8192];
        int read;
        while ((read = in.read(chunk)) != -1) {
            out.write(chunk, 0, read);
        }
        return out.toByteArray();
    }

    // ---- transport -------------------------------------------------------

    @JavascriptInterface
    public void request(final String id, final String command, final String argsJson) {
        executor.execute(new Runnable() {
            @Override
            public void run() {
                JSONObject args;
                try {
                    args = argsJson == null || argsJson.isEmpty()
                            ? new JSONObject()
                            : new JSONObject(argsJson);
                } catch (Exception e) {
                    deliver(id, envelope(command, false, null, "malformed-args", denial(command)));
                    return;
                }
                JSONObject verdict = denial(command);
                if (!verdict.optBoolean("allowed", false)) {
                    deliver(id, envelope(command, false, null, "policy-denied", verdict));
                    return;
                }
                try {
                    Object value = perform(command, args);
                    deliver(id, envelope(command, true, value, null, verdict));
                } catch (Exception e) {
                    deliver(id, envelope(command, false, null, String.valueOf(e.getMessage()), verdict));
                }
            }
        });
    }

    private void deliver(final String id, final JSONObject payload) {
        final String script = "window.__kotobaShellDeliver("
                + JSONObject.quote(id) + "," + JSONObject.quote(payload.toString()) + ")";
        webView.post(new Runnable() {
            @Override
            public void run() {
                webView.evaluateJavascript(script, null);
            }
        });
    }

    // ---- policy ----------------------------------------------------------

    private JSONObject denial(String command) {
        String capability = capabilities.optString(command, null);
        String matchedDeny = firstMatch(deny, command, capability);
        String matchedAllow = firstMatch(allow, command, capability);
        JSONObject verdict = new JSONObject();
        try {
            verdict.put("allowed", matchedDeny == null && matchedAllow != null);
            verdict.put("capability", capability == null ? JSONObject.NULL : capability);
            verdict.put("matched-allow", matchedAllow == null ? JSONObject.NULL : matchedAllow);
            verdict.put("matched-deny", matchedDeny == null ? JSONObject.NULL : matchedDeny);
        } catch (Exception ignored) {
            // JSONObject.put only throws on a null key, which cannot happen here.
        }
        return verdict;
    }

    private static String firstMatch(Set<String> tokens, String command, String capability) {
        if (tokens.contains(command)) {
            return command;
        }
        if (capability != null && tokens.contains(capability)) {
            return capability;
        }
        if (tokens.contains("*")) {
            return "*";
        }
        return null;
    }

    private JSONObject envelope(String command, boolean ok, Object value, String error, JSONObject verdict) {
        JSONObject payload = new JSONObject();
        try {
            payload.put("schema", SCHEMA);
            payload.put("command", command);
            payload.put("ok", ok);
            JSONObject audit = new JSONObject();
            audit.put("audit/schema", "kotoba.shell.audit.v0");
            audit.put("audit/authority", "kotoba-lang/shell");
            audit.put("audit/event", ok ? "provider/execute" : "provider/deny");
            audit.put("audit/target", "android");
            audit.put("audit/command", command);
            audit.put("audit/capability", verdict.opt("capability"));
            audit.put("audit/matched-allow", verdict.opt("matched-allow"));
            audit.put("audit/matched-deny", verdict.opt("matched-deny"));
            payload.put("audit", audit);
            if (value != null) {
                payload.put("value", value);
            }
            if (error != null) {
                payload.put("error", error);
            }
        } catch (Exception ignored) {
            // as above: no null keys are used.
        }
        return payload;
    }

    // ---- providers -------------------------------------------------------

    private Object perform(String command, JSONObject args) throws Exception {
        switch (command) {
            case "clipboard/read-text":
                return object("text", clipboardRead());
            case "clipboard/write-text":
                clipboardWrite(stringArg(args, "text"));
                return object("written", true);
            case "fs/read-text":
                return object("text", fsRead(stringArg(args, "path")));
            case "fs/write-text":
                fsWrite(stringArg(args, "path"), stringArg(args, "text"), false);
                return object("path", args.getString("path"), "written", true);
            case "fs/append-text":
                fsWrite(stringArg(args, "path"), stringArg(args, "text"), true);
                return object("path", args.getString("path"), "appended", true);
            case "keychain/read-text":
                return object("text", keychainRead(stringArg(args, "account")));
            case "keychain/write-text":
                keychainWrite(stringArg(args, "account"), stringArg(args, "text"));
                return object("written", true);
            case "keychain/delete":
                keychainDelete(stringArg(args, "account"));
                return object("deleted", true);
            case "http/fetch":
                return httpFetch(args);
            case "notify/show":
                return notify(stringArg(args, "title"), args.optString("body", ""));
            default:
                throw new IllegalArgumentException("unknown-command: " + command);
        }
    }

    private static String stringArg(JSONObject args, String key) {
        String value = args.optString(key, null);
        if (value == null) {
            throw new IllegalArgumentException("missing string argument: " + key);
        }
        return value;
    }

    private static JSONObject object(Object... pairs) throws Exception {
        JSONObject out = new JSONObject();
        for (int i = 0; i + 1 < pairs.length; i += 2) {
            out.put(String.valueOf(pairs[i]), pairs[i + 1]);
        }
        return out;
    }

    // clipboard: ClipboardManager is main-thread only, and the provider runs on
    // the executor, so both directions hop and wait.

    /**
     * Reads the clipboard, or says why it could not.
     *
     * From Android 10 the system returns nothing to an app that does not hold
     * window focus, and it does so silently: getPrimaryClip returns null and
     * no exception is raised. Returning "" for that would be
     * indistinguishable from an empty clipboard, which is how the first run
     * of this bridge reported a read that the system had actually refused
     * (observed on a booted emulator whose System UI dialog held focus).
     */
    private String clipboardRead() throws Exception {
        final String[] result = new String[]{null};
        final boolean[] focused = new boolean[]{false};
        final boolean[] empty = new boolean[]{false};
        final CountDownLatch latch = new CountDownLatch(1);
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                try {
                    focused[0] = activity.hasWindowFocus();
                    ClipboardManager manager =
                            (ClipboardManager) activity.getSystemService(Context.CLIPBOARD_SERVICE);
                    if (manager == null) {
                        return;
                    }
                    if (!manager.hasPrimaryClip()) {
                        empty[0] = true;
                        return;
                    }
                    ClipData clip = manager.getPrimaryClip();
                    if (clip != null && clip.getItemCount() > 0) {
                        CharSequence text = clip.getItemAt(0).coerceToText(activity);
                        result[0] = text == null ? "" : text.toString();
                    }
                } finally {
                    latch.countDown();
                }
            }
        });
        latch.await();
        if (empty[0]) {
            return "";
        }
        if (result[0] == null) {
            throw new IllegalStateException(focused[0]
                    ? "clipboard read returned nothing while focused"
                    : "clipboard read refused: Android 10+ only allows the focused app to read the clipboard");
        }
        return result[0];
    }

    private void clipboardWrite(final String text) throws Exception {
        final CountDownLatch latch = new CountDownLatch(1);
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                try {
                    ClipboardManager manager =
                            (ClipboardManager) activity.getSystemService(Context.CLIPBOARD_SERVICE);
                    if (manager != null) {
                        manager.setPrimaryClip(ClipData.newPlainText("kotoba-shell", text));
                    }
                } finally {
                    latch.countDown();
                }
            }
        });
        latch.await();
    }

    // fs: scoped to the app's own files dir. Absolute paths and ".." are
    // rejected before resolution, so no spelling reaches the rest of the
    // sandbox.
    private File appDataFile(String path) throws Exception {
        if (path.isEmpty() || path.startsWith("/") || path.contains("..")) {
            throw new IllegalArgumentException("path must be relative and free of '..': " + path);
        }
        File root = new File(activity.getFilesDir(), DATA_DIR);
        File target = new File(root, path);
        File parent = target.getParentFile();
        if (parent != null && !parent.exists() && !parent.mkdirs()) {
            throw new IllegalStateException("could not create " + parent.getPath());
        }
        return target;
    }

    private String fsRead(String path) throws Exception {
        File file = appDataFile(path);
        if (!file.isFile()) {
            throw new IllegalStateException("not readable: " + path);
        }
        return new String(Files.readAllBytes(file.toPath()), StandardCharsets.UTF_8);
    }

    private void fsWrite(String path, String text, boolean append) throws Exception {
        File file = appDataFile(path);
        try (OutputStream out = new FileOutputStream(file, append)) {
            out.write(text.getBytes(StandardCharsets.UTF_8));
        }
    }

    // keychain: Android has no Keychain Services equivalent, so the closest
    // honest mapping is an AndroidKeyStore-held AES-GCM key encrypting values
    // in the app's own SharedPreferences. The key material never leaves the
    // keystore; only iv||ciphertext is stored.
    private SecretKey keychainKey() throws Exception {
        KeyStore keyStore = KeyStore.getInstance("AndroidKeyStore");
        keyStore.load(null);
        if (keyStore.containsAlias(KEY_ALIAS)) {
            KeyStore.SecretKeyEntry entry = (KeyStore.SecretKeyEntry) keyStore.getEntry(KEY_ALIAS, null);
            return entry.getSecretKey();
        }
        KeyGenerator generator = KeyGenerator.getInstance(KeyProperties.KEY_ALGORITHM_AES, "AndroidKeyStore");
        generator.init(new KeyGenParameterSpec.Builder(
                KEY_ALIAS, KeyProperties.PURPOSE_ENCRYPT | KeyProperties.PURPOSE_DECRYPT)
                .setBlockModes(KeyProperties.BLOCK_MODE_GCM)
                .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_NONE)
                .build());
        return generator.generateKey();
    }

    private SharedPreferences keychainPrefs() {
        return activity.getSharedPreferences(KEYCHAIN_PREFS, Context.MODE_PRIVATE);
    }

    private String keychainRead(String account) throws Exception {
        String stored = keychainPrefs().getString(account, null);
        if (stored == null) {
            throw new IllegalStateException("keychain read failed: no item for " + account);
        }
        byte[] raw = Base64.decode(stored, Base64.NO_WRAP);
        byte[] iv = new byte[12];
        System.arraycopy(raw, 0, iv, 0, 12);
        byte[] cipherText = new byte[raw.length - 12];
        System.arraycopy(raw, 12, cipherText, 0, cipherText.length);
        Cipher cipher = Cipher.getInstance("AES/GCM/NoPadding");
        cipher.init(Cipher.DECRYPT_MODE, keychainKey(), new GCMParameterSpec(128, iv));
        return new String(cipher.doFinal(cipherText), StandardCharsets.UTF_8);
    }

    private void keychainWrite(String account, String text) throws Exception {
        Cipher cipher = Cipher.getInstance("AES/GCM/NoPadding");
        cipher.init(Cipher.ENCRYPT_MODE, keychainKey());
        byte[] iv = cipher.getIV();
        byte[] cipherText = cipher.doFinal(text.getBytes(StandardCharsets.UTF_8));
        byte[] joined = new byte[iv.length + cipherText.length];
        System.arraycopy(iv, 0, joined, 0, iv.length);
        System.arraycopy(cipherText, 0, joined, iv.length, cipherText.length);
        keychainPrefs().edit().putString(account, Base64.encodeToString(joined, Base64.NO_WRAP)).commit();
    }

    private void keychainDelete(String account) {
        keychainPrefs().edit().remove(account).commit();
    }

    private JSONObject httpFetch(JSONObject args) throws Exception {
        String urlString = args.optString("url", null);
        if (urlString == null) {
            throw new IllegalArgumentException("missing or malformed url");
        }
        HttpURLConnection connection = (HttpURLConnection) new URL(urlString).openConnection();
        connection.setRequestMethod(args.optString("method", "GET"));
        JSONObject headers = args.optJSONObject("headers");
        if (headers != null) {
            for (Iterator<String> keys = headers.keys(); keys.hasNext(); ) {
                String key = keys.next();
                connection.setRequestProperty(key, headers.optString(key, ""));
            }
        }
        String body = args.optString("body", null);
        if (body != null) {
            connection.setDoOutput(true);
            try (OutputStream out = connection.getOutputStream()) {
                out.write(body.getBytes(StandardCharsets.UTF_8));
            }
        }
        int status = connection.getResponseCode();
        InputStream stream = status >= 400 ? connection.getErrorStream() : connection.getInputStream();
        String text = stream == null ? "" : new String(readAll(stream), StandardCharsets.UTF_8);
        JSONObject responseHeaders = new JSONObject();
        for (Map.Entry<String, java.util.List<String>> entry : connection.getHeaderFields().entrySet()) {
            if (entry.getKey() != null) {
                responseHeaders.put(entry.getKey(), String.join(",", entry.getValue()));
            }
        }
        connection.disconnect();
        return object("status", status, "headers", responseHeaders, "body", text);
    }

    private JSONObject notify(String title, String body) throws Exception {
        NotificationManager manager =
                (NotificationManager) activity.getSystemService(Context.NOTIFICATION_SERVICE);
        if (manager == null) {
            throw new IllegalStateException("no notification manager");
        }
        NotificationChannel channel = new NotificationChannel(
                NOTIFICATION_CHANNEL, "Kotoba Shell", NotificationManager.IMPORTANCE_DEFAULT);
        manager.createNotificationChannel(channel);
        Notification.Builder builder = new Notification.Builder(activity, NOTIFICATION_CHANNEL)
                .setContentTitle(title)
                .setContentText(body)
                .setSmallIcon(android.R.drawable.ic_dialog_info);
        manager.notify((int) (System.nanoTime() & 0x7fffffff), builder.build());
        // Posting succeeds even when the user has denied POST_NOTIFICATIONS;
        // the system drops it silently. Report what was actually done rather
        // than claiming the notification was seen.
        boolean enabled = Build.VERSION.SDK_INT < Build.VERSION_CODES.N
                || manager.areNotificationsEnabled();
        return object("posted", true, "notifications-enabled", enabled);
    }
}

File: resources/kotoba/shell/app/kotoba_shell_bridge.swift
// kotoba-shell in-app provider bridge — the Apple half (iOS and macOS).
//
// Why this file exists: `bin/kotoba-shell-host-ios` runs providers with
// `xcrun simctl spawn booted`, i.e. on the developer's machine against an
// attached simulator. That is a development bridge and none of it exists in a
// build handed to a user. Providers a shipped app can actually call have to be
// compiled into the app, which is what this is.
//
// Policy is enforced here, not only in the CLI. `Resources/
// kotoba-shell-policy.json` is written by `app scaffold` from the same
// `:allow`/`:deny` EDN the CLI evaluates, and the decision below matches
// `kotoba.shell.launcher/policy-decision` clause for clause: a command is
// allowed when no deny token matches and one of {command, capability, "*"} is
// in allow. Deny-by-default: a missing or unreadable policy file allows
// nothing.
//
// {{KEYCHAIN_SERVICE}} / {{DATA_DIR}} are substituted by
// kotoba.shell.native-bridge/apple-bridge-swift.
import Foundation
import Security
import UserNotifications
import WebKit
#if os(iOS)
import UIKit
#elseif os(macOS)
import AppKit
#endif

enum KotobaShellBridgeError: Error {
    case invalidArgument(String)
    case providerFailure(String)
}

final class KotobaShellBridge: NSObject, WKScriptMessageHandlerWithReply {
    static let messageName = "kotobaShell"

    private let allow: Set<String>
    private let deny: Set<String>
    private let capabilities: [String: String]
    private let keychainService: String
    private let dataRoot: URL

    #if os(iOS)
    private static let platformTarget = "ios"
    #else
    private static let platformTarget = "macos"
    #endif

    init(policy: [String: Any]) {
        self.allow = Set((policy["allow"] as? [String]) ?? [])
        self.deny = Set((policy["deny"] as? [String]) ?? [])
        self.capabilities = (policy["capabilities"] as? [String: String]) ?? [:]
        self.keychainService = "{{KEYCHAIN_SERVICE}}"
        let base = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first
            ?? URL(fileURLWithPath: NSTemporaryDirectory())
        self.dataRoot = base.appendingPathComponent("{{DATA_DIR}}", isDirectory: true)
        super.init()
    }

    /// Reads the scaffolded policy. An app whose policy resource is missing
    /// gets an empty allow set rather than an open one -- a packaging mistake
    /// must fail closed, not silently grant every provider.
    static func loadFromBundle() -> KotobaShellBridge {
        guard let url = Bundle.main.url(forResource: "kotoba-shell-policy", withExtension: "json"),
              let data = try? Data(contentsOf: url),
              let parsed = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return KotobaShellBridge(policy: [:])
        }
        return KotobaShellBridge(policy: parsed)
    }

    /// Installs the message handler and the document-start shim on `config`.
    /// Call before creating the WKWebView: userContentController changes made
    /// after a web view exists do not apply to it.
    static func install(into config: WKWebViewConfiguration) -> KotobaShellBridge {
        let bridge = KotobaShellBridge.loadFromBundle()
        config.userContentController.addScriptMessageHandler(
            bridge, contentWorld: .page, name: KotobaShellBridge.messageName)
        if let shim = KotobaShellBridge.shimSource() {
            config.userContentController.addUserScript(
                WKUserScript(source: shim, injectionTime: .atDocumentStart, forMainFrameOnly: true))
        }
        return bridge
    }

    /// Asks for notification authorization once, at launch.
    ///
    /// Not from inside `notify/show`: `requestAuthorization` only calls back
    /// after someone answers the system dialog, so a provider call that
    /// requested inline would block for as long as the dialog stood open and
    /// its result would depend on how fast a person tapped. Android's
    /// MainActivity asks at startup for the same reason. Only asked when the
    /// policy actually allows `notify/show` — an app that never notifies
    /// should not greet its user with a permission sheet.
    func requestNotificationAuthorizationIfAllowed() {
        guard (decision(for: "notify/show")["allowed"] as? Bool) == true,
              Bundle.main.bundleIdentifier != nil else { return }
        UNUserNotificationCenter.current().requestAuthorization(options: [.alert, .sound]) { _, _ in }
    }

    private static func shimSource() -> String? {
        guard let url = Bundle.main.url(forResource: "kotoba-shell-bridge", withExtension: "js") else {
            return nil
        }
        return try? String(contentsOf: url, encoding: .utf8)
    }

    // MARK: - Policy

    private func decision(for command: String) -> [String: Any] {
        let capability = capabilities[command]
        var tokens: [String] = [command, "*"]
        if let capability = capability { tokens.insert(capability, at: 1) }
        let matchedDeny = tokens.first(where: { deny.contains($0) })
        let matchedAllow = tokens.first(where: { allow.contains($0) })
        return [
            "allowed": matchedDeny == nil && matchedAllow != nil,
            "capability": capability as Any,
            "matched-allow": matchedAllow as Any,
            "matched-deny": matchedDeny as Any
        ]
    }

    private func envelope(command: String,
                          ok: Bool,
                          value: Any?,
                          error: String?,
                          decision: [String: Any]) -> [String: Any] {
        var payload: [String: Any] = [
            "schema": "kotoba.shell.bridge.v0",
            "command": command,
            "ok": ok,
            "audit": [
                "audit/schema": "kotoba.shell.audit.v0",
                "audit/authority": "kotoba-lang/shell",
                "audit/event": ok ? "provider/execute" : "provider/deny",
                "audit/target": KotobaShellBridge.platformTarget,
                "audit/command": command,
                "audit/capability": decision["capability"] ?? NSNull(),
                "audit/matched-allow": decision["matched-allow"] ?? NSNull(),
                "audit/matched-deny": decision["matched-deny"] ?? NSNull()
            ]
        ]
        if let value = value { payload["value"] = value }
        if let error = error { payload["error"] = error }
        return payload
    }

    // MARK: - WKScriptMessageHandlerWithReply

    func userContentController(_ userContentController: WKUserContentController,
                               didReceive message: WKScriptMessage,
                               replyHandler: @escaping (Any?, String?) -> Void) {
        guard let body = message.body as? [String: Any],
              let command = body["command"] as? String else {
            replyHandler(nil, "kotoba-shell: request must be {command, args}")
            return
        }
        let args = (body["args"] as? [String: Any]) ?? [:]
        let verdict = decision(for: command)
        guard (verdict["allowed"] as? Bool) == true else {
            replyHandler(envelope(command: command, ok: false, value: nil,
                                  error: "policy-denied", decision: verdict), nil)
            return
        }
        perform(command: command, args: args) { [weak self] result in
            guard let self = self else { return }
            switch result {
            case .success(let value):
                replyHandler(self.envelope(command: command, ok: true, value: value,
                                           error: nil, decision: verdict), nil)
            case .failure(let error):
                replyHandler(self.envelope(command: command, ok: false, value: nil,
                                           error: String(describing: error), decision: verdict), nil)
            }
        }
    }

    // MARK: - Providers

    private func perform(command: String,
                         args: [String: Any],
                         completion: @escaping (Result<Any, Error>) -> Void) {
        do {
            switch command {
            case "clipboard/read-text":
                completion(.success(["text": try clipboardRead()]))
            case "clipboard/write-text":
                try clipboardWrite(try stringArg(args, "text"))
                completion(.success(["written": true]))
            case "fs/read-text":
                completion(.success(["text": try fsRead(try stringArg(args, "path"))]))
            case "fs/write-text":
                let path = try stringArg(args, "path")
                try fsWrite(path, text: try stringArg(args, "text"), append: false)
                completion(.success(["path": path, "written": true]))
            case "fs/append-text":
                let path = try stringArg(args, "path")
                try fsWrite(path, text: try stringArg(args, "text"), append: true)
                completion(.success(["path": path, "appended": true]))
            case "keychain/read-text":
                completion(.success(["text": try keychainRead(try stringArg(args, "account"))]))
            case "keychain/write-text":
                try keychainWrite(try stringArg(args, "account"), text: try stringArg(args, "text"))
                completion(.success(["written": true]))
            case "keychain/delete":
                try keychainDelete(try stringArg(args, "account"))
                completion(.success(["deleted": true]))
            case "http/fetch":
                httpFetch(args, completion: completion)
            case "notify/show":
                notify(title: try stringArg(args, "title"),
                       body: (args["body"] as? String) ?? "",
                       completion: completion)
            default:
                completion(.failure(KotobaShellBridgeError.providerFailure("unknown-command: \(command)")))
            }
        } catch {
            completion(.failure(error))
        }
    }

    private func stringArg(_ args: [String: Any], _ key: String) throws -> String {
        guard let value = args[key] as? String else {
            throw KotobaShellBridgeError.invalidArgument("missing string argument: \(key)")
        }
        return value
    }

    // MARK: clipboard

    private func clipboardRead() throws -> String {
        #if os(iOS)
        return UIPasteboard.general.string ?? ""
        #else
        return NSPasteboard.general.string(forType: .string) ?? ""
        #endif
    }

    private func clipboardWrite(_ text: String) throws {
        #if os(iOS)
        UIPasteboard.general.string = text
        #else
        NSPasteboard.general.clearContents()
        NSPasteboard.general.setString(text, forType: .string)
        #endif
    }

    // MARK: fs (app-data scoped)

    /// Resolves `path` under the app's own data directory. Absolute paths and
    /// `..` are rejected before resolution rather than after, so a caller
    /// cannot reach the rest of the container by any spelling.
    private func appDataURL(_ path: String) throws -> URL {
        if path.hasPrefix("/") || path.contains("..") || path.isEmpty {
            throw KotobaShellBridgeError.invalidArgument("path must be relative and free of '..': \(path)")
        }
        try FileManager.default.createDirectory(at: dataRoot, withIntermediateDirectories: true)
        let url = dataRoot.appendingPathComponent(path)
        let parent = url.deletingLastPathComponent()
        try FileManager.default.createDirectory(at: parent, withIntermediateDirectories: true)
        return url
    }

    private func fsRead(_ path: String) throws -> String {
        let url = try appDataURL(path)
        guard let text = try? String(contentsOf: url, encoding: .utf8) else {
            throw KotobaShellBridgeError.providerFailure("not readable: \(path)")
        }
        return text
    }

    private func fsWrite(_ path: String, text: String, append: Bool) throws {
        let url = try appDataURL(path)
        if append, let handle = try? FileHandle(forWritingTo: url) {
            defer { try? handle.close() }
            try handle.seekToEnd()
            try handle.write(contentsOf: Data(text.utf8))
            return
        }
        try Data(text.utf8).write(to: url, options: .atomic)
    }

    // MARK: keychain

    private func keychainQuery(_ account: String) -> [String: Any] {
        [kSecClass as String: kSecClassGenericPassword,
         kSecAttrService as String: keychainService,
         kSecAttrAccount as String: account]
    }

    private func keychainRead(_ account: String) throws -> String {
        var query = keychainQuery(account)
        query[kSecReturnData as String] = true
        query[kSecMatchLimit as String] = kSecMatchLimitOne
        var item: CFTypeRef?
        let status = SecItemCopyMatching(query as CFDictionary, &item)
        guard status == errSecSuccess, let data = item as? Data,
              let text = String(data: data, encoding: .utf8) else {
            throw KotobaShellBridgeError.providerFailure("keychain read failed: OSStatus \(status)")
        }
        return text
    }

    private func keychainWrite(_ account: String, text: String) throws {
        SecItemDelete(keychainQuery(account) as CFDictionary)
        var query = keychainQuery(account)
        query[kSecValueData as String] = Data(text.utf8)
        query[kSecAttrAccessible as String] = kSecAttrAccessibleAfterFirstUnlock
        let status = SecItemAdd(query as CFDictionary, nil)
        guard status == errSecSuccess else {
            throw KotobaShellBridgeError.providerFailure("keychain write failed: OSStatus \(status)")
        }
    }

    private func keychainDelete(_ account: String) throws {
        let status = SecItemDelete(keychainQuery(account) as CFDictionary)
        guard status == errSecSuccess || status == errSecItemNotFound else {
            throw KotobaShellBridgeError.providerFailure("keychain delete failed: OSStatus \(status)")
        }
    }

    // MARK: http

    private func httpFetch(_ args: [String: Any], completion: @escaping (Result<Any, Error>) -> Void) {
        guard let urlString = args["url"] as? String, let url = URL(string: urlString) else {
            completion(.failure(KotobaShellBridgeError.invalidArgument("missing or malformed url")))
            return
        }
        var request = URLRequest(url: url)
        request.httpMethod = (args["method"] as? String) ?? "GET"
        if let headers = args["headers"] as? [String: String] {
            for (key, value) in headers { request.setValue(value, forHTTPHeaderField: key) }
        }
        if let body = args["body"] as? String { request.httpBody = Data(body.utf8) }
        URLSession.shared.dataTask(with: request) { data, response, error in
            if let error = error {
                completion(.failure(KotobaShellBridgeError.providerFailure(error.localizedDescription)))
                return
            }
            let http = response as? HTTPURLResponse
            var headers: [String: String] = [:]
            for (key, value) in (http?.allHeaderFields ?? [:]) {
                headers[String(describing: key)] = String(describing: value)
            }
            completion(.success([
                "status": http?.statusCode ?? 0,
                "headers": headers,
                "body": String(data: data ?? Data(), encoding: .utf8) ?? ""
            ]))
        }.resume()
    }

    // MARK: notifications

    private func notify(title: String, body: String, completion: @escaping (Result<Any, Error>) -> Void) {
        // UNUserNotificationCenter traps when the process has no bundle
        // identity. A scaffolded app always has one; a host running the same
        // code outside a bundle should get an error, not a crash.
        guard Bundle.main.bundleIdentifier != nil else {
            completion(.failure(KotobaShellBridgeError.providerFailure("notifications require a bundled app")))
            return
        }
        let center = UNUserNotificationCenter.current()
        center.getNotificationSettings { settings in
            // Reads the decision rather than making it. See
            // requestNotificationAuthorizationIfAllowed().
            let authorized = settings.authorizationStatus == .authorized
                || settings.authorizationStatus == .provisional
            guard authorized else {
                completion(.failure(KotobaShellBridgeError.providerFailure(
                    "notification authorization not granted (status \(settings.authorizationStatus.rawValue))")))
                return
            }
            let content = UNMutableNotificationContent()
            content.title = title
            content.body = body
            let request = UNNotificationRequest(identifier: UUID().uuidString, content: content, trigger: nil)
            center.add(request) { addError in
                if let addError = addError {
                    completion(.failure(KotobaShellBridgeError.providerFailure(addError.localizedDescription)))
                } else {
                    completion(.success(["shown": true]))
                }
            }
        }
    }
}
